src/shared/uart_framing/uart_framing.hpp

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <cstdint>
+#include <optional>
+#include <vector>
+
+const uint8_t START_END_FLAG_BYTE = 0x00;
+
+/**
+ * Acts as a frame for uart messages
+ * The type of message should be structs.
+ * These structs should not include pointers or references.
+ *
+ * @tparam T the type of message being framed
+ */
+template <typename T>
+struct UartMessageFrame
+{
+    uint16_t length;
+    uint16_t crc;
+    T message;
+};
+
+namespace
+{
+    /**
+     * Generates a CRC-16 adhering to the CRC-16-CCITT standard
+     *
+     * @param data the data to calculate the crc of
+     * @param length the length of the data
+     * @return CRC-16 for given data and length
+     */
+    template <typename T>
+    uint16_t crc16(const T& data, uint16_t length)
+    {
+        uint8_t x;
+        uint16_t crc = 0xFFFF;
+        auto ptr     = reinterpret_cast<const uint8_t*>(&data);
+        auto bytes   = std::vector<uint8_t>(ptr, ptr + sizeof(data));
```

Comment:
curious why we template here and then cast to an unsigned byte? why not just take an uint8_t pointer to start and leave it up to the user to cast if needed